src/run_graph.rs

```diff
@@ -79,66 +78,106 @@ impl TasksGraph {
         id
     }
 
-    pub fn next_task(&mut self) -> Option<(NodeIndex, Task)> {
+    pub fn next_task<DB: WriteResults>(
+        &mut self,
+        ex: &Experiment,
+        db: &DB,
+    ) -> Option<(NodeIndex, Arc<Task>)> {
         let root = self.root;
-        if let WalkResult::Task(id, task) = self.walk_graph(root) {
+        if let WalkResult::Task(id, task) = self.walk_graph(root, ex, db) {
             Some((id, task))
         } else {
             None
         }
     }
 
-    fn walk_graph(&mut self, node: NodeIndex) -> WalkResult {
+    fn walk_graph<DB: WriteResults>(
+        &mut self,
+        node: NodeIndex,
+        ex: &Experiment,
+        db: &DB,
+    ) -> WalkResult {
+        // Ensure tasks are only executed if needed
+        if let Node::Task {
+            ref task,
+            running: false,
+        } = self.graph[node]
+        {
+            if !task.needs_exec(ex, db) {
+                return WalkResult::NotBlocked;
+            }
+        }
+
         let mut dependencies = 0;
 
         // Try to check for the dependencies of this node
         // The list is collected to make the borrowchecker happy
         let mut neighbors = self.graph.neighbors(node).collect::<Vec<_>>();
         for neighbor in neighbors.drain(..) {
-            match self.walk_graph(neighbor) {
+            match self.walk_graph(neighbor, ex, db) {
                 WalkResult::Task(id, task) => return WalkResult::Task(id, task),
                 WalkResult::Blocked => dependencies += 1,
```

Comment:
We could just early exit with `::Blocked` here I think? And then below we can just assume that there are no dependencies.

Reply:
Well, actually no, if we early return there only one task is executed at the time, because the graph walker has no chance to look for tasks in the other dependent nodes. Fixed in https://github.com/rust-lang-nursery/crater/pull/265